### DISCUSSION TITLE
[Learning] How does the plugin system work?

### DISCUSSION THREAD

**Discussion by Nobody-Special** [dosemu2/dosemu2#2178] [Q&A]
Good mythical rest of the day! :smiley:

After opening a bug report #2177 and concluding that `dlopen` for whatever reason has some issues in sandboxes (Toolbox; Podman) I wanted to investigate why is that.

@stsp pointed that it's likely due to `NOEXEC` on mount of `/dev/shm` which is correct.

Bare-metal mount options:
```sh
user@bare-metal> mount | grep "/dev/shm"
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,seclabel,inode64)
```
Toolbox mount options:
```diff
user@toolbox> mount | grep "/dev/shm"
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,seclabel,inode64)
+tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,seclabel,inode64)
+tmpfs on /run/host/dev/shm type tmpfs (rw,nosuid,nodev,seclabel,inode64)
```

So I have embarked on a mission to learn if there is a possibility to change mount options on the container. (I still have to learn if it is possible through Toolbox alone.) First thing I did was looking for bug reports on Toolbox's Github repo. There I've found containers/toolbox#9 which successfully opens a file descriptor under `/dev/shm/` and the minimal reproducible example present in the above mentioned issue still passes in Toolbox. I've also tested the [minimal example of `dlopen`](https://github.com/fd3kyt/dlopen_example). Of course it also works on Toolbox. If it wouldn't then probably half of the programs would be broken in some way on Toolbox. Hence my question, how does the plugin system work?

**Comment by stsp** [comment]
The toolbox bug report is unrelated.
There was a read-only file system, which
is not `noexec`.
Your minimal dlopen example will not
work if you put your lib into /dev/shm.

**Comment by stsp** [comment]
Also see here:
https://stackoverflow.com/a/76282757
to see why i find it so difficult to find a
(portable) work-around, and here:
https://stackoverflow.com/a/76861853
for a linux-specific work-around.
That work-around uses 2 linux-specific
things: memfd_create() and mmap() via
/proc/self/fd (that one I find a particularly
nasty trick).

See here at the solution that I would
propose:
https://github.com/dosemu2/dosemu2/issues/2159#issuecomment-1959368826
